### PR TITLE
WiimoteReal: Change the scanner to support several backends

### PR DESCRIPTION
**Pull request by leoetlino**
This makes WiimoteScanner support several scanner backends.

This adds a WiimoteScannerBackend base class, which scanner backends derive from, and which allows backend-specific things to be moved out of the common code.

Currently, no platform has more than one scanner backend, but this will probably change with hidapi. (Actually some platforms already sort of have two scanners, such as OS X which has both Bluetooth and HID, and probably Windows too)

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4054)

### CODE REVIEW

**Comment by leoetlino** [comment]
Tested on Linux and Windows, but not on Android and OS X (although it should work).

File: Source/Core/Core/HW/WiimoteReal/WiimoteReal.h [hunk 1]
```diff
@@ -114,6 +114,16 @@ class Wiimote : NonCopyable
   Common::FifoQueue<Report> m_write_reports;
 };
 
+class WiimoteScannerBackend
```
Comment (degasus) [inline Source/Core/Core/HW/WiimoteReal/WiimoteReal.h]: This is not a backend, this is a base class: "WiimoteScannerBase"

Comment (delroth) [inline Source/Core/Core/HW/WiimoteReal/WiimoteReal.h]: I disagree, it's the interface for a backend. The current naming is fine.

**Comment by lioncash** [comment]
Review status: 0 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[Source/Core/Core/HW/WiimoteReal/IOAndroid.h, line 42 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWszNJUM8hLwTxHkkG:-KNWszNJUM8hLwTxHkkH:-1826705332) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOAndroid.h#L42)):*
> ```objc
> {
> public:
>   WiimoteScannerAndroid() {}
> ```

```cpp
WiimoteScannerAndroid() = default;
~WiimoteScannerAndroid() override = default;
```

---

*[Source/Core/Core/HW/WiimoteReal/IOdarwin.h, line 80 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWsSGJNuNa7k63znmn:-KNWsSGJNuNa7k63znmo:-1131792074) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOdarwin.h#L80)):*
> ```objc
> {
> public:
>   WiimoteScannerDarwin() {}
> ```


```cpp
WiimoteScannerDarwin() = default;
~WiimoteScannerDarwin() override = default;
```

---

*[Source/Core/Core/HW/WiimoteReal/IOdarwin.h, line 90 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWrmNr88amzVSM80wG:-KNWrmNsqEEtRFM7ZV-o:-1847700186) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOdarwin.h#L90)):*
> ```objc
> {
> public:
>   WiimoteScannerDarwinHID() {}
> ```

```cpp
WiimoteScannerDarwinHID() = default;
~WiimoteScannerDarwinHID() override = default;
```

---

*[Source/Core/Core/HW/WiimoteReal/IOLinux.h, line 40 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWre7cVU6-Lxp68iR0:-KNWre7cVU6-Lxp68iR1:2047165546) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOLinux.h#L40)):*
> ```objc
> public:
>   WiimoteScannerLinux();
>   ~WiimoteScannerLinux();
> ```

```cpp
~WiimoteScannerLinux() override;
```

---

*[Source/Core/Core/HW/WiimoteReal/IOWin.h, line 48 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWtN7ibfOUl1CjWMln:-KNWtN7jpfFMStJA5-s0:2037343225) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOWin.h#L48)):*
> ```objc
> public:
>   WiimoteScannerWindows();
>   ~WiimoteScannerWindows();
> ```

```cpp
~WiimoteScannerWindows() override;
```

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp, line 470 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWuMf77wLTvEp4pie0:-KNWuMf8iHR6goIInp-2:-2061932912) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp#L470)):*
> ```cpp
> bool WiimoteScanner::IsReady() const
> {
>   if (m_scanner_backends.empty())
> ```

The check for emptiness isn't necessary, `std::any_of` will handle that case. It'll always check if the beginning iterator doesn't match the end iterator before proceeding (in the empty case, this'll be false, and so, false will be returned).

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.h, line 120 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWrC6jLRN47KwoAiBV:-KNWrC6jLRN47KwoAiBW:1309368121) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.h#L120)):*
> ```objc
> {
> public:
>   virtual ~WiimoteScannerBackend(){};
> ```

```cpp
virtual ~WiimoteScannerBackend() = default;
```

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.h, line 137 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWrM5JIc9nzCkl1kRL:-KNWrM5JIc9nzCkl1kRM:2075756060) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.h#L137)):*
> ```objc
> {
> public:
>   WiimoteScanner() {}
> ```

```cpp
WiimoteScanner() = default;
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054)*

**Comment by leoetlino** [comment]
Review status: 0 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[Source/Core/Core/HW/WiimoteReal/IOAndroid.h, line 42 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWszNJUM8hLwTxHkkG:-KNX7O8vxpvLSR1egiKl:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOAndroid.h#L42)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> WiimoteScannerAndroid() = default;
> ~WiimoteScannerAndroid() override = default;
> ```
</details>
Done.

---

*[Source/Core/Core/HW/WiimoteReal/IOdarwin.h, line 80 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWsSGJNuNa7k63znmn:-KNX7Q5Q7tM_91TBLTc1:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOdarwin.h#L80)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> 
> ```cpp
> WiimoteScannerDarwin() = default;
> ~WiimoteScannerDarwin() override = default;
> ```
</details>
Done.

---

*[Source/Core/Core/HW/WiimoteReal/IOdarwin.h, line 90 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWrmNr88amzVSM80wG:-KNX7Qoo7zOrwlmmMsRC:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOdarwin.h#L90)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> WiimoteScannerDarwinHID() = default;
> ~WiimoteScannerDarwinHID() override = default;
> ```
</details>
Done.

---

*[Source/Core/Core/HW/WiimoteReal/IOLinux.h, line 40 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWre7cVU6-Lxp68iR0:-KNX7RsnbdhoGRA-mJNh:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOLinux.h#L40)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> ~WiimoteScannerLinux() override;
> ```
</details>
Done.

---

*[Source/Core/Core/HW/WiimoteReal/IOWin.h, line 48 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWtN7ibfOUl1CjWMln:-KNX7SX042uspUR5Nuzy:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/IOWin.h#L48)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> ~WiimoteScannerWindows() override;
> ```
</details>
Done.

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp, line 470 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWuMf77wLTvEp4pie0:-KNX7V9JvYdzefmE7_If:-1100512967) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp#L470)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> The check for emptiness isn't necessary, `std::any_of` will handle that case. It'll always check if the beginning iterator doesn't match the end iterator before proceeding (in the empty case, this'll be false, and so, false will be returned).
</details>
Done. Forgot to remove this when I changed it to use `std::any_of` instead of `std::all_of`.

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.h, line 120 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWrC6jLRN47KwoAiBV:-KNX7dQ2o30rhVUyFdzL:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.h#L120)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> virtual ~WiimoteScannerBackend() = default;
> ```
</details>
Done.

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.h, line 137 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNWrM5JIc9nzCkl1kRL:-KNX7dxGWxlPh4PDj_vI:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.h#L137)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> WiimoteScanner() = default;
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054)*

**Comment by leoetlino** [comment]
Is there anything else to change? This should lay the groundwork for a future cleanup (by making "disabled" and "emulated" just backends).

**Comment by degasus** [comment]
I dislike the API of the scanner interfaces, but I think the required API cleanup should be within the next PR. So this looks like a good step forward. LGTM

---

Reviewed 7 of 14 files at r1, 1 of 6 files at r3, 5 of 6 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp, line 462 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNrBaBIw4cLjZsCkBGL:-KNrBaBJxyaLQUgpasjK:724544245) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d2976086b6c8d691e5c3090f73001083362f7f67/Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp#L462)):*
> ```cpp
> bool WiimoteScanner::IsReady() const
> {
>   return std::any_of(m_scanner_backends.begin(), m_scanner_backends.end(),
> ```

I'm not so sure if I like this code, but maybe I'm just not familiar with any_of.
IMO this is cleaner:
for (const auto& scanner : m_scanner_backends)
  if (scanner->IsReady())
    return true;
return false;

Feel free to ignore this comment

---

*[Source/Core/Core/HW/WiimoteReal/WiimoteReal.h, line 117 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-KNNqAzhLg8qxUL-MYa9-r2-117:-KNr9SC0sZ1_AZ1xBTvJ:1515371905) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f177da29fb9ba805ce443181953cd01b6852a5bb/Source/Core/Core/HW/WiimoteReal/WiimoteReal.h#L117)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> I disagree, it's the interface for a backend. The current naming is fine.
</details>
But this is not consistent with the other source. But no need to bikeshed about this name...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4054#-:-KNrCmZCPRczR_P9K5Q1:-685932963)*

**Pull request merged**